src/sentry/api/helpers/group_index/index.py

```diff
@@ -275,10 +276,78 @@ def prep_search(
     return result, query_kwargs
 
 
+def serialize_releases(request: Request, group: Group, versions: list[str]) -> list[dict[str, Any]]:
+    releases = {
+        release.version: release
+        for release in Release.objects.filter(
+            projects=group.project,
+            organization_id=group.project.organization_id,
+            version__in=versions,
+        )
+    }
+    serialized_releases = serialize(
+        [releases.get(version) for version in versions],
+        request.user,
+    )
+    # Default to a dictionary if the release object wasn't found and not serialized
+    return [
+        item if item is not None else {"version": version}
+        for item, version in zip(serialized_releases, versions)
+    ]
+
+
+def get_environment_filtered_releases(
+    request: Request, group: Group, environment_names: list[str]
+) -> tuple[dict[str, Any] | None, dict[str, Any] | None]:
+    group_releases = GroupRelease.objects.filter(
+        group_id=group.id,
+        environment__in=environment_names,
+    )
+
+    if not group_releases.exists():
+        return None, None
+
+    first_group_release = group_releases.order_by("first_seen").first()
+    last_group_release = group_releases.order_by("-last_seen").first()
+
+    release_ids = set()
+    if first_group_release:
+        release_ids.add(first_group_release.release_id)
+    if last_group_release:
+        release_ids.add(last_group_release.release_id)
+
+    if not release_ids:
+        return None, None
+
+    releases = {r.id: r for r in Release.objects.filter(id__in=release_ids)}
+
+    first_release = None
+    last_release = None
+
+    if first_group_release and first_group_release.release_id in releases:
+        (first_release,) = serialize_releases(
+            request, group, [releases[first_group_release.release_id].version]
+        )
+
+    if last_group_release and last_group_release.release_id in releases:
+        (last_release,) = serialize_releases(
+            request, group, [releases[last_group_release.release_id].version]
+        )
+
+    return first_release, last_release
+
+
 def get_first_last_release(
     request: Request,
     group: Group,
+    environment_names: list[str] | None = None,
 ) -> tuple[dict[str, Any] | None, dict[str, Any] | None]:
+    if environment_names:
```

Comment:
why can't we just modify `group.get_first_release()` / `group.get_last_release()` to take in the environment names?

Reply:
hmm i'm kind of going back and forth on this. on one hand i agree it would be nice to have a single code path. on the other hand there's caching in `group.get_first_release()` / `group.get_last_release()` which i don't think we would want to rely on if environment filters are active. meaning [here](https://github.com/getsentry/sentry/blob/7cbdda03defbe2d2869b4a316de61673e7e62963/src/sentry/models/release.py#L159), i don't think we want to start caching release versions for various environment combinations, which means we'd have to check the cache only if the environment filter is not being used. but maybe that's fine

Reply:
@JoshFerge could you let me know what you think of this updated pattern?